File: algorithms/leet.0230.src.1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    var idx = 0, result = null;
    var walk = function(ptr) {
        if(result !== null) return;
        if(ptr.left) walk(ptr.left);
        if(++idx == k) result = ptr.val;
        if(ptr.right) walk(ptr.right);
    };
    walk(root);
    return result;
};

File: algorithms/leet.0230.src.1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def dfs(nd):
            nonlocal k
            if not nd: return None
            a = dfs(nd.left)
            if a is not None: return a
            k -= 1
            if not k: return nd.val
            return dfs(nd.right)
        return dfs(root)
